lib/ansible/modules/yum_repository.py

```diff
@@ -614,6 +631,59 @@ def main():
     if module.params['params']:
         module.fail_json(msg="The params option to yum_repository was removed in Ansible 2.5 since it circumvents Ansible's option handling")
 
+    if module.params['name'] is None and module.params['copr'] is None:
+        module.fail_json(msg="Either the 'name' or the 'copr' parameter need to be set")
+
+    if module.params['copr']:
+        copr_api = module.params['copr_api']
+        copr_user, copr_project = module.params['copr'].split('/')
+        copr_server = generic_urlparse(list(urlparse(copr_api)))['hostname']
+        distro = get_distribution().lower()
+        if distro in ['centos', 'rhel', 'redhat']:
+            distro = 'epel'
+
+        # RHEL/CentOS will give a version like '7.8' or '8.2'
+        # but Fedora will give a version like '31' or '32' or 'Rawhide'
+        if distro == 'epel':
+            distro_version = get_distribution_version().split('.')[0]
+        else:
+            distro_version = get_distribution_version().lower()
+        copr_url = '{0}/coprs/{1}/{2}/repo/{3}-{4}/dnf.repo'.format(copr_api, copr_user, copr_project, distro, distro_version)
+        response, info = fetch_url(module, copr_url)
+        if info['status'] != 200:
+            module.fail_json(msg="failed to fetch COPR repo {0}, error was: {1}".format(copr_url, info['msg']))
+
+        copr_params = configparser.RawConfigParser()
+
+        # Python 2 ConfigParser doesn't have read_string and readfp is
+        # deprecated in Python 3
+        try:
+            copr_params.read_string(to_text(response.read()))
+        except AttributeError:
+            copr_params.readfp(StringIO(to_text(response.read())))
+
+        # Take all the parameters from the yum repo we just got from the COPR
+        # server and use them to set module.params for this module. We only set
+        # the variable if it hasn't been set by the user already.
+        #
+        # What this means in practical terms is the user can specify 'baseurl:
+        # http://example.com/packages/' in their playbook and that will override
+        # the baseurl we get from the COPR server.
+        copr_server_user_project = "copr:{0}:{1}:{2}".format(copr_server, copr_user, copr_project)
+        for key in dict(copr_params.items(copr_server_user_project)).keys():
+            if not module.params.get(key, None):
+                module.params[key] = copr_params.get(copr_server_user_project, key)
+
+        module.params['name'] = copr_server_user_project
+        module.params['description'] = copr_params.get(copr_server_user_project, 'name')
+        module.params['file'] = "_{0}".format(copr_server_user_project)
```

Comment:
Is it worth warning if the user provides these to the module and they are about to get forcibly changed?

Reply:
I've moved the code so now the user can override it.